shared/src/main/scala/mlscript/TypeSimplifier.scala

```diff
@@ -361,14 +366,24 @@ trait TypeSimplifier { self: Typer =>
                   tts.toArray.sorted // TODO also filter out tts that are inherited by the class
                     .foldLeft(withType: ST)(_ & _)
                 case _ =>
-                  lazy val newRcd = rcd.copy(
-                    rcd.fields.filterNot(_._1.name.isCapitalized).mapValues(go(_, pol)))(rcd.prov).sorted
-                  LhsRefined(bo.map {
-                    case ct: ClassTag => ct
-                    case ft @ FunctionType(l, r) => FunctionType(go(l, !pol), go(r, pol))(ft.prov)
-                    case tt @ TupleType(fs) => TupleType(fs.mapValues(go(_, pol)))(tt.prov)
-                    case wt @ Without(b, ns) => Without(go(b, pol), ns)(wt.prov)
-                  }, tts, newRcd).toType(sort = true)
+                  val nFields = rcd.fields.filterNot(_._1.name.isCapitalized).mapValues(go(_, pol))
+                  val (res, nfs) = bo match {
+                    case S(tt @ TupleType(fs)) => 
```

Comment:
The implementation here is too crude. Consider this:

```scala
def test: (string, 1) & { _1: "hello" }
//│ test: (string, 1,) & {_1: "hello", _2: 1}
```

It should instead result in `("hello", 1,)`.

You need to merge all tuple fields into the tuple, and to only leave non-tuple fields in the refinement!

It's okay though. I suggest to open a new issue to improve this later, and merge this PR without the improvement.

Reply:
Actually I just added a comment in https://github.com/hkust-taco/mlscript/issues/69#issuecomment-1031281164

Reply:
Please feel free to close this conversation when you've read it.